synapse/rest/client/v1/room.py

```diff
@@ -267,6 +284,108 @@ def on_PUT(self, request, room_id, event_type, txn_id):
         )
 
 
+class RoomBulkSendEventRestServlet(TransactionRestServlet):
+    def __init__(self, hs):
+        super().__init__(hs)
+        self.event_creation_handler = hs.get_event_creation_handler()
+        self.auth = hs.get_auth()
+
+        self._msc2716_enabled = hs.config.experimental.msc2716_enabled
+
+    def register(self, http_server):
+        # /rooms/$roomid/bulksend
+        PATTERNS = "/rooms/(?P<room_id>[^/]*)/bulksend"
+        register_txn_path(self, PATTERNS, http_server, with_get=True)
+
+    async def on_POST(self, request, room_id):
+        requester = await self.auth.get_user_by_req(request, allow_guest=False)
+        body = parse_json_object_from_request(request)
+        assert_params_in_dict(body, ["state_events_at_start", "events"])
+
+        prev_events_from_query = parse_strings_from_args(request.args, "prev_event")
+
+        logger.info("body waewefaew %s", body)
+
+        auth_event_ids = []
+        for stateEv in body["state_events_at_start"]:
+            logger.info("stateEv %s", stateEv)
+            assert_params_in_dict(stateEv, ["type", "content", "sender"])
+
+            event_dict = {
+                "type": stateEv["type"],
+                "content": stateEv["content"],
+                "room_id": room_id,
+                "sender": stateEv["sender"],  # requester.user.to_string(),
+                "state_key": stateEv["state_key"],
+            }
+
+            # Make the events float off in their own
+            fake_prev_event_id = "$" + random_string(43)
+
+            # Also mark the event as an outlier outside of the normal DAG
+            (event, _,) = await self.event_creation_handler.create_event(
+                requester, event_dict, prev_event_ids=[fake_prev_event_id], outlier=True
```

Comment:
> To mark an event as an outlier we set `event.internal_metadata.outlier = False`
>
> https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r603461607

But I assume you meant `event.internal_metadata.outlier = True`?

Reply:
I'm an idiot :facepalm: